core/src/test/scala/org/apache/spark/ui/memory/MemoryListenerSuite.scala

```diff
@@ -0,0 +1,258 @@
+import org.apache.spark.scheduler.cluster._
+
+class MemoryListenerSuite extends SparkFunSuite with LocalSparkContext {
+  test("test HashMap size for MemoryListener") {
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+    val execId2 = "exec-2"
+
+    (1 to 2).foreach { i =>
+      listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(i))
+      listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(i))
+    }
+    // stages are all completed, no activeStages now
+    assert(listener.activeStagesToMem.isEmpty)
+
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, new ExecutorMetrics))
+    // ExecutorMetrics is not related with Stages directly
+    assert(listener.activeStagesToMem.isEmpty)
+
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(3))
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId2, new ExecutorMetrics))
+    // totally 2 executors updated their metrics
+    assert(listener.activeExecutorIdToMem.size === 2)
+    assert(listener.activeStagesToMem.size === 1)
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(3))
+
+    assert(listener.activeStagesToMem.isEmpty)
+    assert(listener.completedStagesToMem.size === 3)
+    assert(listener.activeExecutorIdToMem.size === listener.latestExecIdToExecMetrics.size)
+    assert(listener.removedExecutorIdToMem.isEmpty)
+  }
+
+  test("test first stage with no executor metrics update") {
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId1, new ExecutorInfo("host1", 1, Map.empty)))
+
+    // stage 1, no metrics update
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(1))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(1))
+
+    // stage 2, with one metrics update
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(2))
+    val execMetrics = MemoryListenerSuite.createExecutorMetrics("host-1", 0L, 20, 10)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, execMetrics))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(2))
+
+    val mapForStage1 = listener.completedStagesToMem.get((1, 0)).get
+    // no metrics for stage 1 since no metrics update for stage 1
+    assert(mapForStage1.get(execId1).get.transportInfo === None)
+    val mapForStage2 = listener.completedStagesToMem.get((2, 0)).get
+    assert(mapForStage2.size === 1)
+    val memInfo = mapForStage2.get(execId1).get
+    assert(memInfo.transportInfo.isDefined)
+    val transMetrics = memInfo.transportInfo.get
+    assert((20, 10, MemTime(20, 0), MemTime(10, 0)) === (transMetrics.onHeapSize,
+      transMetrics.offHeapSize, transMetrics.peakOnHeapSizeTime, transMetrics.peakOffHeapSizeTime))
```

Comment:
or since you do a similar assert a lot, use a helper method which takes the 4 expected values and a `transMetrics`, and does the 4 asserts.